substratevm/src/com.oracle.svm.reflect/src/com/oracle/svm/reflect/dynamicclassloading/hosted/DynamicClassLoadingFeature.java

```diff
@@ -0,0 +1,168 @@
+import com.oracle.svm.core.util.UserError;
+import com.oracle.svm.hosted.FeatureImpl;
+import com.oracle.svm.hosted.ImageClassLoader;
+import com.oracle.svm.hosted.config.ConfigurationParserUtils;
+import com.oracle.svm.reflect.MultiClassLoaderReporter;
+import com.oracle.svm.util.ReflectionUtil;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.hosted.Feature;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@AutomaticFeature
+public class DynamicClassLoadingFeature implements Feature {
+
+    class DynamicClassContainer {
+        Class<?> clazz;
+        String name;
+        String checksum;
+
+        DynamicClassContainer(Class<?> clazz, String name, String checksum) {
+            this.clazz = clazz;
+            this.name = name;
+            this.checksum = checksum;
+        }
+    }
+
+    private Map<String, DynamicClassContainer> dynamicClassContainers = new HashMap<>();
+    private List<StringBuilder> errMsgs = new ArrayList<>();
+
+    /**
+     * Reflectively call protected method URLClassLoader.addURL.
+     *
+     * @param url url to add into the URLClassLoader
+     * @param classLoader URLClassLoader instance
+     */
+    private static void addURL2URLClassLoader(URL url, URLClassLoader classLoader) {
+        Method addURL = ReflectionUtil.lookupMethod(URLClassLoader.class, "addURL", URL.class);
+        try {
+            addURL.invoke(classLoader, url);
+        } catch (IllegalAccessException e) {
+            e.printStackTrace();
+        } catch (InvocationTargetException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Reflectively call protected method URLClassLoader.findClass.
+     *
+     * @param name target class name
+     * @param classLoader
+     * @return the class found with the given name
+     * @throws ClassNotFoundException
+     */
+    private static Class<?> callFindClass(String name, URLClassLoader classLoader) throws ClassNotFoundException {
+        Method findClass = ReflectionUtil.lookupMethod(URLClassLoader.class, "findClass", String.class);
+        try {
+            return (Class<?>) findClass.invoke(classLoader, name);
+        } catch (IllegalAccessException | InvocationTargetException e) {
+            throw new ClassNotFoundException(name);
+        }
+    }
+
+    @Override
+    public void afterRegistration(AfterRegistrationAccess a) {
+        FeatureImpl.AfterRegistrationAccessImpl access = (FeatureImpl.AfterRegistrationAccessImpl) a;
+        ImageClassLoader imageClassLoader = access.getImageClassLoader();
+        URLClassLoader cl = (URLClassLoader) imageClassLoader.getClassLoader();
```

Comment:
Assuming that the image class loader is a `URLClassLoader` and manipulating it through reflection does not seem like a good idea at all (especially with the module system). These steps need proper support in `ImageClassLoader` and `AbstractNativeImageClassLoaderSupport` and its subclasses and that probably entail creating a `ClassLoader` subclass for these purposes.

Moreover it seems very roundabout and not safe to reconstruct a class directory hierarchy, add it to the loader's class path (which potentially) and load the class from there. Is it not possible to just load the class from a byte array with the file's contents (other than that we need to call `defineClass` somehow)?

This _really_ needs fixing.

Reply:
Now read the class file into byte array and then reflectively call `defineClass` with the byte array data.